### DISCUSSION TITLE
building nedit-ng for CentOS 7

### DISCUSSION THREAD

**Discussion by Necoboss** [eteran/nedit-ng#342] [General]
Hi Guys,

Has anyone managed to build nedit-ng on CentOS 7?

I've been trying for a few days now without success so far :'(

Any tips or help would be pretty much appreciated.

Thanks and Regards,
Necoboss

**Comment by eteran** [comment]
I haven't tried that specific OS yet, but it's easy enough to spin up a container and see what the story is.

Can you share some of the errors you've seen?

**Comment by Necoboss** [comment]
Hi eteran!

First, I tried without root access, and here what I've got:

cd /phyd/users/cdoming
git clone https://github.com/eteran/nedit-ng.git 
cd nedit-ng/
mkdir DEPENDENCIES-LOCAL-INSTALL
cd DEPENDENCIES-LOCAL-INSTALL
   <install locally boost/QT5/cmake>
    some dependencies where installed but cmake could not find it so I have to try it locally
    as an example boost was already installed however I didn't manage to make cmake find it
    $ yum list installed | grep boost
    boost.x86_64                            1.53.0-28.el7                  @base    

cd ..
mkdir build
cd build

# qt5 local path setting
export Qt5_DIR=/gpfs/phyd/users/cdoming/nedit-ng/DEPENDENCIES-LOCAL-INSTALL/QT5/5.12.12/gcc_64

# using specific version of cmake
../DEPENDENCIES-LOCAL-INSTALL/cmake-3.25.2-linux-x86_64/bin/cmake .. -DBOOST_ROOT=/phyd/users/cdoming/nedit-ng/DEPENDENCIES-LOCAL-INSTALL/boost_1_55_0

-- The CXX compiler identification is GNU 4.8.5
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/lib64/ccache/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Boost: /phyd/users/cdoming/nedit-ng/DEPENDENCIES-LOCAL-INSTALL/boost_1_55_0 (found suitable version "1.55.0", minimum required is "1.35")  
-- Extracting nedit-ng version info with git...
-- Found Git: /tools_hpc/tools/utils/git/2.33.0/bin/git (found version "2.33.0") 
-- NEDIT_VERSION_GIT = "2020.1-251-g79989ea6"
-- NEDIT_COMMIT_GIT = "79989ea68a229568a133167c71d6847ad10ba532"
-- NEDIT_BRANCH_GIT = "master"
-- Found X11: /usr/include   
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Configuring done
-- Generating done
-- Build files have been written to: /phyd/users/cdoming/nedit-ng/build


 make
[  1%] Building CXX object Regex/CMakeFiles/Regex.dir/Execute.cpp.o
c++: error: unrecognized command line option \u2018-Wsuggest-final-methods\u2019
c++: error: unrecognized command line option \u2018-Wsuggest-final-types\u2019
c++: error: unrecognized command line option \u2018-Wsuggest-override\u2019
make[2]: *** [Regex/CMakeFiles/Regex.dir/Execute.cpp.o] Error 1
make[1]: *** [Regex/CMakeFiles/Regex.dir/all] Error 2
make: *** [all] Error 2

Any clue of what could be wrong here? gcc version? boost install?

Thanks and regards,
NecoBoss

**Comment by anjohnson** [comment]
I've been building nedit-ng on RHEL-7 for many years. You will need updated versions of binutils, gcc and g++, and probably Cmake, Bison and Boost too to build it with — check the minimum versions that the nedit-ng documentation specifies. On the official Red Hat release I use their devtoolset-7 with packages like these to provide GCC:

- devtoolset-7-binutils-2.28-11.el7.x86_64
- devtoolset-7-gcc-7.3.1-5.16.el7.x86_64
- devtoolset-7-gcc-c++-7.3.1-5.16.el7.x86_64

I assume Centos 7 provide equivalents to them; newer versions would also probably work.

This is the Makefile I add to the top-level of the nedit-ng working directory to configure and build it to run from my personal home directory:

```makefile
# My build configuration

PREFIX = /home/anjohnson

TOOLSET = /opt/rh/devtoolset-7
CMAKE = /opt/cmake/bin/cmake
BUILD = Build

CMAKE_TARGETS = all install

CMAKE_OPTIONS += -DCMAKE_INSTALL_PREFIX=$(PREFIX)

# UI preference configuration:
CMAKE_OPTIONS += -DPER_TAB_CLOSE=ON

default: install

# NB: $(BUILD) is a real directory
$(BUILD):
	mkdir $@

$(BUILD)/Makefile: $(BUILD)
	cd $(BUILD) && \
	source $(TOOLSET)/enable && \
	$(CMAKE) $(CMAKE_OPTIONS) ..

$(CMAKE_TARGETS): %: $(BUILD)/Makefile
	$(MAKE) -C $(BUILD) -j $@

clean:
	test -f $(BUILD)/Makefile && \
	$(MAKE) -C $(BUILD) $@

distclean:
	rm -rf $(BUILD)

.PHONY: $(CMAKE_TARGETS) clean distclean
.NOTPARALLEL:
```

**Comment by eteran** [comment]
Thanks @anjohnson !

Saving me a lot of time with this helpful build tips :-)

**Comment by Necoboss** [comment]
Thanks @anjohnson !

I verified the following: 
```
$ yum list installed |grep devtoolset-7
devtoolset-7-binutils.x86_64            2.28-11.el7                    @centos-sclo-rh
devtoolset-7-gcc.x86_64                 7.3.1-5.16.el7                 @centos-sclo-rh
devtoolset-7-gcc-c++.x86_64             7.3.1-5.16.el7                 @centos-sclo-rh
devtoolset-7-gcc-gdb-plugin.x86_64      7.3.1-5.16.el7                 @centos-sclo-rh
devtoolset-7-gcc-gfortran.x86_64        7.3.1-5.16.el7                 @centos-sclo-rh
devtoolset-7-gcc-plugin-devel.x86_64    7.3.1-5.16.el7                 @centos-sclo-rh
devtoolset-7-libquadmath-devel.x86_64   7.3.1-5.16.el7                 @centos-sclo-rh
devtoolset-7-libstdc++-devel.x86_64     7.3.1-5.16.el7                 @centos-sclo-rh
devtoolset-7-runtime.x86_64             7.1-4.el7                      @centos-sclo-rh
```

and after doing : 
$ scl enable devtoolset-7 bash

I was able to compile nedit-ng ! (so it was clearly a gcc / c++ version issue)

However I don't understand why I had to install locally boost and give its path while boost with a compatible version is available

```
$ yum list installed | grep boost
boost.x86_64                            1.53.0-28.el7                  @base    
boost-atomic.x86_64                     1.53.0-28.el7                  @base    
boost-chrono.x86_64                     1.53.0-28.el7                  @base    
boost-context.x86_64                    1.53.0-28.el7                  @base    
boost-date-time.x86_64                  1.53.0-28.el7                  @base    
boost-filesystem.x86_64                 1.53.0-28.el7                  @base    
boost-graph.x86_64                      1.53.0-28.el7                  @base    
boost-iostreams.x86_64                  1.53.0-28.el7                  @base    
boost-locale.x86_64                     1.53.0-28.el7                  @base    
boost-math.x86_64                       1.53.0-28.el7                  @base    
boost-program-options.x86_64            1.53.0-28.el7                  @base    
boost-python.x86_64                     1.53.0-28.el7                  @base    
boost-random.x86_64                     1.53.0-28.el7                  @base    
boost-regex.x86_64                      1.53.0-28.el7                  @base    
boost-serialization.x86_64              1.53.0-28.el7                  @base    
boost-signals.x86_64                    1.53.0-28.el7                  @base    
boost-system.x86_64                     1.53.0-28.el7                  @base    
boost-test.x86_64                       1.53.0-28.el7                  @base    
boost-thread.x86_64                     1.53.0-28.el7                  @base    
boost-timer.x86_64                      1.53.0-28.el7                  @base    
boost-wave.x86_64                       1.53.0-28.el7                  @base    
```

Any idea why ?

Regards,
NecoBoss

PS: I had to do this: 
`cmake .. -DBOOST_ROOT=/gpfs/tools_hpc/tools/utils/nedit/2020.1/build/boost_1_81_0`

**Reply by eteran** [reply]
My guess is you had to install the boost devel packages because they provide the headers but the ones you had installed are library only.

**Reply by anjohnson** [reply]
That makes sense, @Necoboss it looks like you didn't have the boost-devel RPM installed, and that's essential. My RHEL-7 box has `boost-devel-1.53.0-28.el7.x86_64` which includes the headers for some if not all of the RPMs that you list as having installed.

**Reply by Necoboss** [reply]
Thank you guys for the explanations.

As both of you mentioned, I understood I had to install boost and boost-devel packages
`$ yum install -y boost`
and 
`$ yum install -y boost-devel`

```
$ yum list installed  | grep -P "^boost\.|^boost-devel"
boost.x86_64                            1.53.0-28.el7                  @base    
boost-devel.x86_64                      1.53.0-28.el7                  @base 
```

And it worked for boost! 
This might help others!

Thanks and regards,
NecoBoss

PS: Now no need to -DBOOST_ROOT option while invoking cmake
